Add storage for the starting corner ID

Adimote opened the pull request:
See the [design doc](https://docs.google.com/document/d/1UcmNURVxKhkmlk0yM_ABMjMIcFJv_3TqbyJ5AIUnNeo/edit?usp=sharing) for this feature.

Add new type of board; the startup board, which is automatically
inserted on startup.

Add new board, 'GameState' which stores any values it is
received, and reports them when a status command is called.

prophile commented on robotd/devices.py:
```diff
@@ -97,6 +97,30 @@ def status(self):
 
+class GameState(Board):
+    """
+    State storage for the game. keeps a store of whet it is sent
```
Typos. This could also collapse down to one line.

Adimote replied:
Fixed

prophile commented:
@Adimote could you cherry-pick the unrelated changes to master, or pull them out to a separate PR? This is hard to review with unrelated changes :)

prophile commented on robotd/devices.py:
```diff
@@ -40,7 +40,7 @@ def make_safe(self):
         This is called after control connections have died.
         """
-        # Brake both the motors
+        # set both motors to brake
```
I'm generally in favour of comments being capitalised.

prophile replied:
Also why is this change in this commit?

prophile commented on robotd/devices.py:
```diff
@@ -97,6 +97,30 @@ def status(self):
         return {'temperature': temp_milli_degrees / 1000}
 
 
+class GameState(Board):
```
I'm not sure about the semantic overload of `Board` here. Could we do something else or rename `Board` and clarify what it covers?

Adimote replied:
I agree with this, however I think it should be fixed after the release next Monday

prophile replied:
I feel like this is mostly a matter of finding a better name than `Board` (which was somewhat misleading with devices like `BrainTemperatureSensor` anyway) and doing a find-and-replace—it needn't be a big drama and can happen soon :)

Adimote replied:
Will rename the `Board` in a later commit

prophile commented on robotd/devices.py:
```diff
@@ -97,6 +97,30 @@ def status(self):
+
+    def __init__(self):
+        super().__init__({})
+        self._state = {'corner': 0, 'mode': 'development'}
```
This could be made "public": state isn't private and it might be convenient to access (e.g. in tests).

prophile replied:
Do we want to distinguish between development and competition mode?

prophile replied:
And should we be calling it 'corner' rather than, say, 'zone'? It does mean changing the API if robots start not in corners in a future game.

Adimote replied:
1. done
2. We'll need to keep the state of competition at some point if we want to have a 3 minute shut-off timer.
3. The 'zone' comment is fair, will fix

prophile commented on robotd/devices.py:
```diff
@@ -171,7 +195,7 @@ def __init__(self, node, camera=None, token_sizes=None):
         self.thread = None
         self.vision = self._create_vision(camera)
-        self.running = False
+        self.stop_event = Event()
```
In future it would be easier to review if commits were broken out a bit more :)

prophile commented on robotd/devices_base.py:
```diff
@@ -11,12 +11,15 @@ class BoardMeta(type):
         cls = type.__new__(cls, name, bases, dict)
         if hasattr(cls, 'lookup_keys') and cls.enabled:
             BoardMeta.BOARDS.append(cls)
+        if hasattr(cls, 'create_on_startup'):
```
This should probably check for `enabled`—and I think `create_on_startup` should probably be defined in the base class since it's a flag so that we can just do `if cls.create_on_startup`. Also avoids the slightly surprising result that `create_on_startup = False` still results in the board going into `STARTUP_BOARDS`!

Adimote replied:
fixed

prophile commented on robotd/devices_base.py:
```diff
@@ -11,12 +11,15 @@ class BoardMeta(type):
         if hasattr(cls, 'lookup_keys') and cls.enabled:
             BoardMeta.BOARDS.append(cls)
+        if hasattr(cls, 'create_on_startup'):
+            BoardMeta.STARTUP_BOARDS.append(cls)
```
Is it better to do this rather than leaving it to users of `BOARDS` to check the flag when iterating over it?

Adimote replied:
fixed

prophile commented on robotd/master.py:
```diff
@@ -11,12 +11,23 @@
+    # Split the message into buf_size chunks
+    buf_size = BoardRunner.SOCK_BUFFER_SIZE
+    for msg in [message[i:i + buf_size] for i in range(0, len(message), buf_size)]:
+        connection.send(msg)
```
This isn't valid on a `SOCK_SEQPACKET` socket. Messages are atomic. We're using `SEQPACKET` precisely because this is guaranteed—it's one message per read() call.

Adimote replied:
The issue is the vision code produces Large packets, I wasn't happy with receiving 524288 byte packets just to make sure it can handle the worst possible case of the camera seeing all 512 markers. Plus this number would need to be arbitrarily raised if we added more info to the vision API. It felt better to have a dynamic packet size

prophile replied:
Dynamic packet size isn't a thing—one message is one message. If we can't read a dynamically-sized message from `SOCK_SEQPACKET` then we should drop it altogether and use `SOCK_STREAM` detecting end-of-message with a `\n`.

Adimote replied:
I disagree with your original statement, it's **up to** one message per read() call, see the Socket API Spec text about SOCK_SEQPACKET:

> A record can be sent using one or more output operations and received using one or more input operations, but a single operation never transfers parts of more than one record.

source: http://pubs.opengroup.org/onlinepubs/009695399/functions/xsh_chap02_10.html

Adimote replied:
Therefore I posit that it's fine to split records up like they have been here

prophile replied:
I stand corrected.

Given those semantics I'm not clear on why we need to use multiple send calls?

Adimote replied:
I'm trying to avoid calling recv(524288) as that's a big buffer size. From my testing, if you ever call recv() on something smaller than the size of the sent packet, data is lost. Therefore we need to enforce a maximum packet size on the sending end

Adimote replied:
this change is irrelevant to the corner_id branch, I'll move it to another PR

prophile commented on robotd/master.py:
```diff
@@ -11,12 +11,23 @@
 class BoardRunner(multiprocessing.Process):
     """Control process for one board."""
 
+    SOCK_BUFFER_SIZE = 2049
```
Why this number?

Adimote replied:
typo, will fix

Adimote committed:
Add storage for the starting corner ID

See the design doc for this feature.

Add new type of board; the startup board, which is automatically
inserted on startup.

Add new board, 'GameState' which stores any values it is
received, and reports them when a status command is called.

Adimote committed:
whitespace

Adimote committed:
make GameState.state public

Adimote committed:
rename 'corner' to 'zone' in api
